Fix 2513, fix canonical_channel_name when platform is appended

mcg1969 opened the pull request:
Motivated by #2513. While investigating that issue, I discovered that `canonical_channel_name` would not return the correct result if the platform suffix (e.g., `/noarch`, `/osx-64`) was included in the URL, so I fixed that as well.

The primary fix for #2513 is to have `fetch_index` call `normalize_urls` if the caller passes in a bare list/iterable of channels. This converts the list into the format needed later in `fetch_index` which is a dictionary: the keys are the long channel URLs, and the keys are (canonical channel name, priority) tuples.

Removed some unused options in `canonical_channel_name` as well.

mcg1969 commented:
@ilanschnell if there's a way to test this branch with constructor before we merge, I'd appreciate it.

ilanschnell commented:
Hmm, just tried it, but now I get a different error (which I don't get using conda 4.0.6):
```
Traceback (most recent call last):
  File "/home/ilan/a/bin/constructor", line 6, in <module>
    sys.exit(main())
  File "/home/ilan/constructor/constructor/main.py", line 146, in main
    verbose=opts.verbose)
  File "/home/ilan/constructor/constructor/main.py", line 81, in main_build
    fcp.main(info, verbose=verbose)
  File "/home/ilan/constructor/constructor/fcp.py", line 170, in main
    for url in info['channels']))
  File "/home/ilan/conda/conda/fetch.py", line 263, in fetch_index
    for url in iterkeys(channel_urls)]
  File "/home/ilan/conda/conda/fetch.py", line 68, in func
    res = f(*args, **kwargs)
  File "/home/ilan/conda/conda/fetch.py", line 150, in fetch_repodata
    raise RuntimeError(msg)
RuntimeError: Could not find URL: http://repo.continuum.io/pkgs/free/linux-64
```

mcg1969 commented:
Thanks! That's certainly why I asked. I just wasn't confident the tests were covering your case. I'll continue to investigate

mcg1969 commented:
Yeah, `fetch_repodata` is very picky about seeing a trailing slash on those URLs. The normalization code I added wasn't doing that in some cases. I've corrected that.

msarahan commented:
I don't think it is fetch_repodata that is picky.  Repo does strange
redirects to port 8080 when the trailing slash is not supplied.

On Thu, May 19, 2016, 18:25 Michael C. Grant <notifications@github.com>
wrote:

> Yeah, fetch_repodata is very picky about seeing a trailing slash on those
> URLs. The normalization code I added wasn't doing that in some cases. I've
> corrected that.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/conda/conda/pull/2518#issuecomment-220475222>
>

ilanschnell commented:
This work now, but I have a new bug.  When I tell constructor to build a `linux-64` installer (on my `osx-64` system, i.e.:
```
$ constructor examples/grin/ --platform linux-64
```
which works fine with `conda 4.0.6`, I now get:
```
Traceback (most recent call last):
  File "/Users/ilan/python/bin/constructor", line 6, in <module>
    exec(compile(open(__file__).read(), __file__, 'exec'))
  File "/Users/ilan/constructor/bin/constructor", line 6, in <module>
    sys.exit(main())
  File "/Users/ilan/constructor/constructor/main.py", line 146, in main
    verbose=opts.verbose)
  File "/Users/ilan/constructor/constructor/main.py", line 81, in main_build
    fcp.main(info, verbose=verbose)
  File "/Users/ilan/constructor/constructor/fcp.py", line 170, in main
    for url in info['channels']))
  File "/Users/ilan/conda/conda/fetch.py", line 269, in fetch_index
    repodatas = [(u, f.result()) for u, f in zip(urls, futures)]
  File "/Users/ilan/python/lib/python2.7/site-packages/concurrent/futures/_base.py", line 403, in result
    return self.__get_result()
  File "/Users/ilan/python/lib/python2.7/site-packages/concurrent/futures/thread.py", line 55, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/ilan/conda/conda/fetch.py", line 68, in func
    res = f(*args, **kwargs)
  File "/Users/ilan/conda/conda/fetch.py", line 150, in fetch_repodata
    raise RuntimeError(msg)
RuntimeError: Could not find URL: http://repo.continuum.io/pkgs/free/linux-64/osx-64/
```

ilanschnell commented:
As a side note.  Constructor only uses `fetch_index` and `fetch_pkg` from `conda.fetch`.

mcg1969 commented:
I already tried to fix this custom `platform` case. Well, it looks like I'm going to have to pull down the `constructor` code and figure this out.

ilanschnell commented:
Or, you can look the code in `conda 4.0.6`, and restore the behavior.  The constructor code is (where it calls the fetch fetch functions) is in those three places:
  * https://github.com/conda/constructor/blob/master/constructor/fcp.py#L147
  * https://github.com/conda/constructor/blob/master/constructor/fcp.py#L162
  * https://github.com/conda/constructor/blob/master/constructor/fcp.py#L168

mcg1969 commented:
> Or, you can look the code in conda 4.0.6, and restore the behavior.

We're not going to be rolling back channel prioritization, Ilan. I'll find a solution here obviously, but going back to 4.0.6 is not an option. My latest commit here is an attempt to factor out the new behavior though.

ilanschnell commented:
I didn't say that we should restore 4.0.6, just the *behavior* of the fetch functions.

mcg1969 commented:
Oh, yes, absolutely. I definitely agree that `fetch_index` should continue to support the calling sequence it had at 4.0.6

mcg1969 commented:
In fact @kalefranz will be happy to know that this fix is allowing us to remove the hack he had to put into `conda info` ;-)

ilanschnell commented:
Great, it all works now.

mcg1969 commented:
Excellent. Thanks for checking, and sorry for the trouble!

mcg1969 commented:
@kalefranz, let's merge!
